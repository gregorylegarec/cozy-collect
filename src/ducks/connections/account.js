import {
  CREATE_CONNECTION,
  UPDATE_CONNECTION_ERROR,
  UPDATE_CONNECTION_RUNNING_STATUS
} from './'

const reducer = (state = {}, action) => {
  switch (action.type) {
    case CREATE_CONNECTION:
      return state
    case UPDATE_CONNECTION_ERROR:
      const { error } = action
      if (error) return { ...state, error }
      return state
    case UPDATE_CONNECTION_RUNNING_STATUS:
      const { isRunning } = action
      const hasRun = state.hasRun || !!state.isRunning && !isRunning
      return { ...state, isRunning, hasRun }
    default:
      return state
  }
}

export default reducer

// selectors
export const isRunning = (state) => {
  return state.isRunning
}

export const hasRun = (state) => {
  return state.hasRun
}
