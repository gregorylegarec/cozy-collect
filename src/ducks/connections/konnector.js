import {
  CREATE_CONNECTION,
  UPDATE_CONNECTION_ERROR,
  UPDATE_CONNECTION_RUNNING_STATUS
} from './'

import account, { isRunning, hasRun } from './account'

const reducer = (state = {}, action) => {
  switch (action.type) {
    case CREATE_CONNECTION:
    case UPDATE_CONNECTION_ERROR:
    case UPDATE_CONNECTION_RUNNING_STATUS:
      return { ...state, [action.account._id]: account(state[action.account._id], action) }
    default:
      return state
  }
}

export default reducer

// selectors
export const hasRunningConnection = (state) => {
  return Object.keys(state).find(key => isRunning(state[key]))
}

export const hasRunConnection = (state) => {
  return Object.keys(state).find(key => hasRun(state[key]))
}
