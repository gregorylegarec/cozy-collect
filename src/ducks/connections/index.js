import konnector, {
  getConnectionStatus,
  hasError,
  hasRunningConnection,
  hasRunConnection
} from './konnector'

// constant
export const CREATE_CONNECTION = 'CREATE_CONNECTION'
export const UPDATE_CONNECTION_RUNNING_STATUS = 'UPDATE_CONNECTION_RUNNING_STATUS'
export const UPDATE_CONNECTION_ERROR = 'UPDATE_CONNECTION_ERROR'

// reducers

const reducer = (state = {}, action) => {
  switch (action.type) {
    case CREATE_CONNECTION:
    case UPDATE_CONNECTION_ERROR:
    case UPDATE_CONNECTION_RUNNING_STATUS:
      if (!action.konnector || !action.konnector.slug) throw new Error('Missing konnector slug')
      return { ...state, [action.konnector.slug]: konnector(state[action.konnector.slug], action) }
    default:
      return state
  }
}

export default reducer

// action creators sync
export const createConnection = (konnector, account, folder) => ({
  type: CREATE_CONNECTION,
  konnector,
  account,
  folder
})

export const updateConnectionError = (konnector, account, error) => ({
  type: UPDATE_CONNECTION_ERROR,
  konnector,
  account,
  error
})

export const updateConnectionRunningStatus = (konnector, account, isRunning = false) => ({
  type: UPDATE_CONNECTION_RUNNING_STATUS,
  konnector,
  account,
  isRunning
})

// selectors
const getKonnectorIconURL = (registry, slug) => {
  const url = `assets/icons/konnectors/${slug}.svg`
  let icon = null
  try {
    icon = require(url)
  } catch (error) {
    console.warn(`Cannot get icon ${url}: ${error.message}`)
  }
  return icon
}

const queueStatuses = {
  done: 'loaded',
  error: 'failed',
  running: 'loading'
}

export const getQueue = (state, konnectorsRegistry) => {
  return Object.keys(state).reduce((runningConnections, key) => {
    const konnector = state[key]
    const label = konnectorsRegistry[key] && konnectorsRegistry[key].name
    const connectionStatus = getConnectionStatus(state[key])
    const status = queueStatuses[connectionStatus] || connectionStatus
    const icon = getKonnectorIconURL(konnectorsRegistry, key)
    if (hasRunningConnection(konnector) || hasRunConnection(konnector)) {
      runningConnections.push({
        label,
        status,
        icon
      })
    }
    return runningConnections
  }, [])
}

export const getRun = (state) => {
  return Object.keys(state).reduce((numRunConnections, slug) => {
    const konnector = state[slug]
    return (hasRunConnection(konnector) && !hasRunningConnection(konnector))
      ? numRunConnections + 1
        : numRunConnections
  }, 0)
}

export const getSuccessfulRun = (state) => {
  return Object.keys(state).reduce((numRunConnections, slug) => {
    const konnector = state[slug]
    return (hasRunConnection(konnector) && !hasError(konnector))
      ? numRunConnections + 1
        : numRunConnections
  }, 0)
}
