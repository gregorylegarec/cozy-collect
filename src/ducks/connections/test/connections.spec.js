/* eslint-env jest */

import connections, {
   createConnection,
   updateConnectionError,
   updateConnectionRunningStatus
 } from '../'

describe('Connections Duck', () => {
  describe('createConnection', () => {
    it('adds new connection in empty state', () => {
      const state = undefined
      const konnector = { slug: 'cozy' }
      const account = { _id: '9bf93550308311c59f0a0047fc00fa1b' }

      const result = connections(state, createConnection(konnector, account))

      expect(result).toMatchSnapshot()
    })

    it('adds new connection', () => {
      const state = {
        testprovider: {
          '17375ac5a59e4d6585fc7d1e1c75ec74': {}
        }
      }
      const konnector = { slug: 'cozy' }
      const account = { _id: '9bf93550308311c59f0a0047fc00fa1b' }

      const result = connections(state, createConnection(konnector, account))

      expect(result).toMatchSnapshot()
    })
  })

  describe('updateConnectionError', () => {
    it('set an error', () => {
      const state = {
        testprovider: {
          '17375ac5a59e4d6585fc7d1e1c75ec74': {}
        }
      }
      const konnector = { slug: 'testprovider' }
      const account = { _id: '17375ac5a59e4d6585fc7d1e1c75ec74' }
      const error = new Error('test error')

      const result = connections(state, updateConnectionError(konnector, account, error))

      expect(result).toMatchSnapshot()
    })
  })

  describe('updateConnectionRunningStatus', () => {
    it('set a connection to idle', () => {
      const state = {
        testprovider: {
          '17375ac5a59e4d6585fc7d1e1c75ec74': {}
        }
      }
      const konnector = { slug: 'testprovider' }
      const account = { _id: '17375ac5a59e4d6585fc7d1e1c75ec74' }

      const result = connections(state, updateConnectionRunningStatus(konnector, account, false))

      expect(result).toMatchSnapshot()
    })

    it('set a connection to running', () => {
      const state = {
        testprovider: {
          '17375ac5a59e4d6585fc7d1e1c75ec74': {}
        }
      }
      const konnector = { slug: 'testprovider' }
      const account = { _id: '17375ac5a59e4d6585fc7d1e1c75ec74' }

      const result = connections(state, updateConnectionRunningStatus(konnector, account, true))

      expect(result).toMatchSnapshot()
    })

    it('set `hasRun` to true when setting `isRunning` from true to false', () => {
      const state = {
        testprovider: {
          '17375ac5a59e4d6585fc7d1e1c75ec74': {
            isRunning: true
          }
        }
      }
      const konnector = { slug: 'testprovider' }
      const account = { _id: '17375ac5a59e4d6585fc7d1e1c75ec74' }

      const result = connections(state, updateConnectionRunningStatus(konnector, account, false))

      expect(result).toMatchSnapshot()
    })

    it('keeps `hasRun` to true after setting `isRunning` from false to true', () => {
      const state = {
        testprovider: {
          '17375ac5a59e4d6585fc7d1e1c75ec74': {
            isRunning: false,
            hasRun: true
          }
        }
      }
      const konnector = { slug: 'testprovider' }
      const account = { _id: '17375ac5a59e4d6585fc7d1e1c75ec74' }

      const result = connections(state, updateConnectionRunningStatus(konnector, account, true))

      expect(result).toMatchSnapshot()
    })
  })
})
