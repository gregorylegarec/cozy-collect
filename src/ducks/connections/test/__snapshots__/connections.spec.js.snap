exports[`Connections Duck createConnection adds new connection 1`] = `
Object {
  "cozy": Object {
    "9bf93550308311c59f0a0047fc00fa1b": Object {},
  },
  "testprovider": Object {
    "17375ac5a59e4d6585fc7d1e1c75ec74": Object {},
  },
}
`;

exports[`Connections Duck createConnection adds new connection in empty state 1`] = `
Object {
  "cozy": Object {
    "9bf93550308311c59f0a0047fc00fa1b": Object {},
  },
}
`;

exports[`Connections Duck updateConnectionError set an error 1`] = `
Object {
  "testprovider": Object {
    "17375ac5a59e4d6585fc7d1e1c75ec74": Object {
      "error": [Error: test error],
    },
  },
}
`;

exports[`Connections Duck updateConnectionRunningStatus keeps \`hasRun\` to true after setting \`isRunning\` from false to true 1`] = `
Object {
  "testprovider": Object {
    "17375ac5a59e4d6585fc7d1e1c75ec74": Object {
      "hasRun": true,
      "isRunning": true,
    },
  },
}
`;

exports[`Connections Duck updateConnectionRunningStatus set \`hasRun\` to true when setting \`isRunning\` from true to false 1`] = `
Object {
  "testprovider": Object {
    "17375ac5a59e4d6585fc7d1e1c75ec74": Object {
      "hasRun": true,
      "isRunning": false,
    },
  },
}
`;

exports[`Connections Duck updateConnectionRunningStatus set a connection to idle 1`] = `
Object {
  "testprovider": Object {
    "17375ac5a59e4d6585fc7d1e1c75ec74": Object {
      "hasRun": false,
      "isRunning": false,
    },
  },
}
`;

exports[`Connections Duck updateConnectionRunningStatus set a connection to running 1`] = `
Object {
  "testprovider": Object {
    "17375ac5a59e4d6585fc7d1e1c75ec74": Object {
      "hasRun": false,
      "isRunning": true,
    },
  },
}
`;
